tools/analysis_tools/analyze_results.py

```diff
@@ -35,62 +39,48 @@ def parse_args():
+    vis.dataset_meta = {'classes': dataset.CLASSES}
 
     # save imgs
-    show_keys = ['pred_score', 'pred_class', 'gt_class']
     for result in results:
-        result_show = dict((k, v) for k, v in result.items() if k in show_keys)
-        outfile = osp.join(full_dir, osp.basename(result['filename']))
-        model.show_result(result['filename'], result_show, out_file=outfile)
+        data_sample = ClsDataSample().set_gt_label(
+            result['gt_label']).set_pred_label(
+                result['pred_label']).set_pred_score(
+                    torch.Tensor(result['pred_scores']))
+        img = mmcv.imread(result['img_path'], channel_order='rgb')
+        vis.add_datasample(result['filename'], img, data_sample)
+
+    mmengine.dump(results, osp.join(full_dir, folder_name + '.json'))
 
 
 def main():
     args = parse_args()
 
     # load test results
-    outputs = mmcv.load(args.result)
-    assert ('pred_score' in outputs and 'pred_class' in outputs
-            and 'pred_label' in outputs), \
-        'No "pred_label", "pred_score" or "pred_class" in result file, ' \
-        'please set "--out-items" in test.py'
+    outputs = mmengine.load(args.result)
 
-    cfg = mmcv.Config.fromfile(args.config)
+    cfg = mmengine.Config.fromfile(args.config)
     if args.cfg_options is not None:
         cfg.merge_from_dict(args.cfg_options)
 
-    model = build_classifier(cfg.model)
-
     # build the dataloader
-    dataset = build_dataset(cfg.data.test)
-    filenames = list()
-    for info in dataset.data_infos:
-        if info['img_prefix'] is not None:
-            filename = osp.join(info['img_prefix'],
-                                info['img_info']['filename'])
-        else:
-            filename = info['img_info']['filename']
-        filenames.append(filename)
-    gt_labels = list(dataset.get_gt_labels())
-    gt_classes = [dataset.CLASSES[x] for x in gt_labels]
-
-    outputs['filename'] = filenames
-    outputs['gt_label'] = gt_labels
-    outputs['gt_class'] = gt_classes
-
-    need_keys = [
-        'filename', 'gt_label', 'gt_class', 'pred_score', 'pred_label',
-        'pred_class'
-    ]
-    outputs = {k: v for k, v in outputs.items() if k in need_keys}
+    cfg.test_dataloader.dataset.pipeline = []
+    dataset = build_dataset(cfg.test_dataloader.dataset)
+
     outputs_list = list()
-    for i in range(len(gt_labels)):
+    for i in range(len(outputs)):
         output = dict()
-        for k in outputs.keys():
-            output[k] = outputs[k][i]
+        output['img_path'] = outputs[i]['img_path']
+        output['filename'] = Path(outputs[i]['img_path']).name
+        output['gt_label'] = int(outputs[i]['gt_label']['label'][0])
+        output['pred_score'] = float(
+            torch.max(outputs[i]['pred_label']['score']).item())
+        output['pred_scores'] = outputs[i]['pred_label']['score'].tolist()
+        output['pred_label'] = int(outputs[i]['pred_label']['label'][0])
```

Comment:
Not all datasets have the `img_path` key. For example, the CIFAR dataset directly loads all images during initialization and only has `img` key.
Better to use `sample_idx` to get the data info from the dataset directly.

```suggestion
        output['sample_idx'] = outputs[i]['sample_idx']
        output['gt_label'] = outputs[i]['gt_label']['label']
        output['pred_score'] = float(
            torch.max(outputs[i]['pred_label']['score']).item())
        output['pred_scores'] = outputs[i]['pred_label']['score']
        output['pred_label'] = outputs[i]['pred_label']['label']
```

And in the `save_imgs`:
```python
    for result in results:
        data_sample = ClsDataSample()\
            .set_gt_label(result['gt_label'])\
            .set_pred_label(result['pred_label'])\
            .set_pred_score(result['pred_scores'])
        data_info = dataset.get_data_info(result['sample_idx'])
        if 'img' in data_info:
            img = data_info['img']
            name = str(result['sample_idx'])
        elif 'img_path' in data_info:
            img = mmcv.imread(data_info['img_path'], channel_order='rgb')
            name = Path(data_info['img_path']).name
        else:
            raise ValueError('Cannot load images from the dataset infos.')
        if rescale_factor is not None:
            img = mmcv.imrescale(img, rescale_factor)
        vis.add_datasample(name, img, data_sample)

        for k, v in result.items():
            if isinstance(v, torch.Tensor):
                result[k] = v.tolist()
```